File: elixir/test/decrypt_entity_test.exs
# Decrypt entity test (offline, mock transport)

defmodule BluefinDecryptx.DecryptEntityTest do
  use ExUnit.Case

  alias Voxgig.Struct, as: S
  alias BluefinDecryptx.Helpers, as: H
  alias BluefinDecryptx.Json

  defp fixture do
    Json.parse(File.read!("../.sdk/test/entity/decrypt/DecryptTestData.json"))
  end

  defp mk_sdk do
    existing = H.or_(S.getpath(fixture(), "existing"), S.jm([]))
    BluefinDecryptx.test(S.jm(["entity", existing]))
  end

  defp first_id do
    existing = H.or_(S.getpath(fixture(), "existing.decrypt"), S.jm([]))
    keys = S.keysof(existing)
    if keys == [], do: nil, else: hd(keys)
  end

  test "should create instance" do
    sdk = BluefinDecryptx.test()
    ent = BluefinDecryptx.decrypt(sdk)
    assert ent != nil
  end

  test "should list records" do
    sdk = mk_sdk()
    ent = BluefinDecryptx.decrypt(sdk)
    result = BluefinDecryptx.Entity.Decrypt.list(ent, S.jm([]))
    assert S.islist(result)
  end

  test "should create then read back" do
    sdk = BluefinDecryptx.test(S.jm(["entity", S.jm(["decrypt", S.jm([])])]))
    ent = BluefinDecryptx.decrypt(sdk)
    made = BluefinDecryptx.Entity.Decrypt.create(ent, S.jm(["name", "test-create"]))
    assert S.ismap(made)
    assert S.getprop(made, "id") != nil
  end
end

File: scala/sdktest/DecryptEntityTest.scala
// Generated basic-flow test for the decrypt entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped DecryptTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefindecryptxsdk.core.{Helpers, BluefinDecryptxSDK}
import voxgig.bluefindecryptxsdk.utility.struct.Struct

object DecryptEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("decrypt.instance") {
      val testsdk = BluefinDecryptxSDK.testSDK()
      val ent = testsdk.decrypt(null)
      rep.check("decrypt.instance", ent != null, "expected non-null decrypt entity")
    }

    rep.scope("decrypt.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/decrypt/DecryptTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinDecryptxSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("decrypt01", "DECRYPT01")
      idmap.put("decrypt02", "DECRYPT02")
      idmap.put("decrypt03", "DECRYPT03")
      val now = System.currentTimeMillis()

      // CREATE
      val decryptRef01Ent = client.decrypt(null)
      var decryptRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.decrypt"), "decrypt_ref01"))
      val decryptRef01DataResult = decryptRef01Ent.create(decryptRef01Data, null)
      decryptRef01Data = Helpers.toMapAny(decryptRef01DataResult)
      rep.check("decrypt.create.map", decryptRef01Data != null, "expected create result to be a map")

      // LIST
      val decryptRef01Match = new LinkedHashMap[String, Object]()
      val decryptRef01ListResult = decryptRef01Ent.list(decryptRef01Match, null)
      rep.check("decrypt.list.islist", decryptRef01ListResult.isInstanceOf[JList[?]], "expected list result to be an array, got " + decryptRef01ListResult)
    }
  }
}
